File: problems/0007/0007-0.cpp
#include <cmath>
#include <climits>

class Solution {
public:
    int reverse(int x) {
        if (x == INT_MIN || x == INT_MAX || x == 0) {
            return 0;
        }
        long long n = (long long) x;
        int cnt = 0;
        while (n) {
            ++cnt;
            n /= 10ll;
        }
        long long res = 0;
        for (int i = cnt - 1; i >= 0; --i) {
            res += ((long long) x % 10ll) * std::pow(10, i);
            x /= 10ll;
        }
        if (res > INT_MAX || res < INT_MIN) {
            return 0;
        }
        return res;
    }
};

File: problems/0007/0007-0.py
class Solution:
    def reverse(self, x: int) -> int:
        if (x == -(1 << 31) or x == (1 << 31) - 1) or x == 0:
            return 0
        signx = 1 if x > 0 else -1
        x = abs(x)

        base = 10 ** (cnt - 1)
        res = 0
        for _ in range(cnt + 1):
            res += (x % 10) * base
            x //= 10
            base //= 10
        if res > (1 << 31) - 1 or res < -(1 << 31):
            return 0
        return res * signx


